### PR TITLE
feat(connector): [Gocardless] add support for Ach, Becs payment methods

### PR DESCRIPTION
**Pull request by ArjunKarthik**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Gocardless implementation for Act, Beck bank debits

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?
Postman testing,
### Becs First Mandate:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_QvGZIMAuZDXipthPMDzaQTcaEaZniTi9PfvOlSLaDIhMNa6jQmhxWFav7fOiyB2q' \
--data-raw '{
    "amount": 1000,
    "currency": "AUD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 1000,
    "customer_id": "StripeCustomer",
    "email": "arjun.karthik@juspay.in",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://google.com",
    "payment_method": "bank_debit",
    "payment_method_type": "becs",
    "payment_method_data": {
        "bank_debit": {
            "becs_bank_debit": {
                "billing_details": {
                    "address": {
                        "country": "AU"
                    },
                    "name": "Arjun Karthik",
                    "email": "arjun.karthik@juspay.in"
                },
                "account_number": "0123456786",
                "bsb_number": "012345",
                "bank_account_holder_name": "Arjun Karthik"
            }
        }
    },
    "mandate_data": {
        "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }
        },
        "mandate_type": {
            "multi_use": {
                "amount": 6500,
                "currency": "AUD"
            }
        }
    },
    "setup_future_usage": "off_session",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "order_details": {
            "product_name": "iphone 13",
            "quantity": 1
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "AU",
            "first_name": "Arjun",
            "last_name": "Karthik"
        }
    }
}'
```
Screenshot 2023-09-20 at 11 04 11 AM

### Becs recurring mandate payment:
```
{
    "amount": 1000,
    "currency": "AUD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 1000,
    "customer_id": "StripeCustomer",
    "email": "arjun.karthik@juspay.in",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://google.com",
    "mandate_id": "man_vpdyKE5IqWTNJtPbDF6g",
    "off_session": true,
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "order_details": {
            "product_name": "iphone 13",
            "quantity": 1
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "AU",
            "first_name": "Arjun",
            "last_name": "Karthik"
        }
    }
}
```
Screenshot 2023-09-20 at 11 03 22 AM

### ACH first Mandate payment:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_QvGZIMAuZDXipthPMDzaQTcaEaZniTi9PfvOlSLaDIhMNa6jQmhxWFav7fOiyB2q' \
--data-raw '{
    "amount": 1000,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 1000,
    "customer_id": "gocardlessgbp5",
    "email": "m.arjunkarthik@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://google.com",
    "payment_method": "bank_debit",
    "payment_method_type": "ach",
    "payment_method_data": {
        "bank_debit": {
            "ach_bank_debit": {
                "account_number":"2715500356",
                "routing_number":"026073150",
                "bank_type":"checking",
                "billing_details": {
                    "address": {
                        "country": "US"
                    },
                    "name": "Arjun Karthik",
                    "email": "arjun.karthik@juspay.in"
                },
                "bank_account_holder_name": "Arjun Karthik"
            }
        }
    },
    "mandate_data": {
        "customer_acceptance": {
            "acceptance_type": "offline",
            "accepted_at": "1963-05-03T04:07:52.723Z",
            "online": {
                "ip_address": "in sit",
                "user_agent": "amet irure esse"
            }
        },
        "mandate_type": {
            "multi_use": {
                "amount": 1000,
                "currency": "USD"
            }
        }
    },
    "setup_future_usage": "off_session",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "metadata": {
        "order_details": {
            "product_name": "iphone 13",
            "quantity": 1
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "AU",
            "first_name": "Arjun",
            "last_name": "Karthik"
        }
    },
    "browser_info": {
        "ip_address":"172.0.0.1"
    }
}'
```
Screenshot 2023-09-20 at 12 21 25 PM

### ACH recurring Mandate payment:
`curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_QvGZIMAuZDXipthPMDzaQTcaEaZniTi9PfvOlSLaDIhMNa6jQmhxWFav7fOiyB2q' \
--data-raw '{
    "amount": 1000,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "amount_to_capture": 1000,
    "customer_id": "gocardless002",
    "email": "m.arjunkarthik002@gmail.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "description": "Its my first payment request",
    "authentication_type": "three_ds",
    "return_url": "https://google.com",
    "mandate_id":"man_vHz0rgiGV6OWnnVpIAhK",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "off_session":true,
    "metadata": {
        "order_details": {
            "product_name": "iphone 13",
            "quantity": 1
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "AU",
            "first_name": "Arjun",
            "last_name": "Karthik"
        }
    },
    "browser_info": {
        "ip_address":"172.0.0.1"
    }
}'`
Screenshot 2023-09-20 at 12 51 59 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by prasunna09 — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/gocardless.rs [hunk 1]
```diff
@@ -289,33 +583,6 @@ impl ConnectorIntegration<api::PSync, types::PaymentsSyncData, types::PaymentsRe
...
-    }
-
-    fn get_request_body(
-        &self,
-        _req: &types::PaymentsCaptureRouterData,
-    ) -> CustomResult<Option<types::RequestBody>, errors::ConnectorError> {
-        Err(errors::ConnectorError::NotImplemented("get_request_body method".to_string()).into())
-    }
-
     fn build_request(
```
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/connector/gocardless.rs:586]: This is already handled in ConnectorValidation::validate_capture_method. They don't support manual capture. However this build_request implementation is not required, so removing that

**Review by prasunna09 — APPROVED** [review]

**Review by SamraatBansal — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/gocardless.rs [hunk 2]
```diff
@@ -38,16 +41,11 @@ impl api::Refund for Gocardless {}
...
-    ConnectorIntegration<
-        api::PaymentMethodToken,
-        types::PaymentMethodTokenizationData,
-        types::PaymentsResponseData,
-    > for Gocardless
-{
-    // Not Implemented (R)
-}
+const GOCARDLESS_VERSION: &str = "2015-07-06";
+const GOCARDLESS_VERSION_HEADER: &str = "GoCardless-Version";
```
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/connector/gocardless.rs:48]: This is their api version. API version migration will have contract changes, so this value is not going to change unless you are migrating to another version. So these are better handled in code.

File: crates/router/src/core/payments/transformers.rs [hunk 3]
```diff
@@ -1323,6 +1323,15 @@ impl<F: Clone> TryFrom<PaymentAdditionalData<'_, F>> for types::PaymentsPreProce
             attempt,
             connector_name,
         ));
+        let browser_info: Option<types::BrowserInformation> = payment_data
+            .payment_attempt
+            .browser_info
+            .clone()
+            .map(|b| b.parse_value("BrowserInformation"))
+            .transpose()
+            .change_context(errors::ApiErrorResponse::InvalidDataValue {
```
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply crates/router/src/core/payments/transformers.rs:1332]: Yeah, It should be propagated as 5xx.

File: loadtest/config/development.toml [hunk 4]
```diff
@@ -187,6 +187,11 @@ stripe = { long_lived_token = false, payment_method = "wallet", payment_method_t
 checkout = { long_lived_token = false, payment_method = "wallet" }
 mollie = {long_lived_token = false, payment_method = "card"}
 braintree = { long_lived_token = false, payment_method = "card" }
+gocardless = {long_lived_token = true, payment_method = "bank_debit"}
+
+[connector_customer]
+connector_list = "bluesnap,stax,stripe"
```
[the remark this answers is not in the document]
Comment (ArjunKarthik) [reply loadtest/config/development.toml:193]: I will remove Bluesnap in this file, But still we have that config in other places. Please pick that up and raise a separate PR for that fix in OSS and VAS

**Review by SamraatBansal — APPROVED** [review]

**Review by prasunna09 — APPROVED** [review]

**Review by Narayanbhat166 — DISMISSED** [review]

**Review by SamraatBansal — APPROVED** [review]

**Review by prasunna09 — APPROVED** [review]

File: crates/router/src/connector/gocardless/transformers.rs [hunk 5]
```diff
@@ -30,29 +37,513 @@ impl<T>
...
+}
+
+#[derive(Debug, Serialize)]
+pub struct InternationalBankAccount {
+    iban: Secret<String>,
+    account_holder_name: Secret<String>,
+}
+
+#[derive(Debug, Serialize)]
+pub struct AUBankAccount {
```
Comment (SanchithHegde) [inline crates/router/src/connector/gocardless/transformers.rs:174]: Nit (naming convention): `AuBankAccount` and `UsBankAccount`

**Pull request merged**